Revise consistent snapshots sections (#54)
secure-systems-lab/peps: pep-0458.txt

__ https://en.wikipedia.org/wiki/Collision_(computer_science)

> Clients, such as pip, using the TUF protocol MUST be modified to download

Reviewer: ```suggestion
Clients, such as pip,  that use the TUF protocol MUST be modified to download
```

> every metadata and target file (except for *timestamp* metadata) by including,

Reviewer: ```suggestion
every metadata and target file, except for *timestamp* metadata. This is done  by including
```

Producing Consistent Snapshots
------------------------------

When a new project release is uploaded to PyPI, PyPI MUST update the *bin-n*

> metadata responsible for the target files of the project release (remember that

Reviewer: Take the parentheses out.
```suggestion
metadata responsible for the target files of the project release. Remember that 
```

Reviewer: Take out the closed parens as well.

target files are sorted into bins by their filename hashes). Consequentially,
PyPI MUST update *snaphot* to account for the updated *bin-n* metadata, and
*timestamp* to account for the updated *snapshot* metadata. These updates
SHOULD be handled by automated processes, e.g. one or more *transaction

> processes* and one *snapshot process*.

Reviewer 2: Can you define the transaction processes and explain how many new targets should be included in each process?

> The transaction process keeps track of a project upload, adds all new target

Reviewer 2: ```suggestion
Each transaction process keeps track of a project upload, adds all new target
```

Reviewer: Agree with Marina's edit

> files to the most recent, relevant *bin-n* metadata (in memory) and informs the

Reviewer 3: You don't need to keep bin-n metadata in RAM all the time, but this is likely the most efficient way to do it: keep it in RAM, and occasionally write copies to disk.

Author: We don't have to be specific about it. My thinking was that the transaction process should not make it available for download just yet (that's what the snapshot process should do).

snapshot process to produce a consistent snapshot. Each project release SHOULD
be handled in an atomic transaction, so that a given consistent snapshot
contains all target files of a project release. However, transaction processes

> MAY be parallelized under the following constraints:

Reviewer 3: The problem with parallelization is that even if two different transaction processes work on different projects, they need to be aware of the version number increments to the bin-n roles. Safest to just work on an event-by-event concurrency model.

I see that this is clarified later: that snapshot metadata is updated serially instead of bin-n metadata, but this needs to be clearer upfront.

Author: I thought the purpose of separating transaction processes from the snapshot process was to parallelize?

> - No pair of transaction processes MUST concurrently work on the same project.

Reviewer 2: I think the MUST here is a bit confusing. maybe something like "Each pair MUST NOT work concurrently"

Reviewer: How about this?
```suggestion
- Pairs of transaction processes MUST NOT  work concurrently on the same project.
```

- No pair of transaction processes MUST concurrently work on projects that
  belong to the same *bin-n* role.

When a transaction process is finished updating the relevant *bin-n* metadata
it informs the snapshot process to generate a new consistent snapshot. The
snapshot process does so by taking the updated *bin-n* metadata, incrementing
their respective version numbers, signing them with the *bin-n* role key(s),
and writing them to *VERSION_NUMBER.bin-N.json*.

Similarly, the snapshot process then takes the most recent *snapshot* metadata,

> updates its *bin-n* metadata hashes and version numbers, increments its own

Author: Just realized that hashes are optional, and we also leave it out of the overhead calcs. Should I remove it? Put it in parentheses?

Reviewer 2: For simplicity I'd say remove it

Reviewer 3: Yes, remove the hashes, please

Author: will do

version number, signs it with the *snapshot* role key, and writes it to
*VERSION_NUMBER.snapshot.json*.

And finally, the snapshot process takes the most recent *timestamp* metadata,
updates its *snapshot* metadata hash and version number, increments its own
version number, sets a new expiration time, signs it with the *timestamp* role
key, and writes it to *timestamp.json*.

The snapshot process MUST generate consistent snapshots sequentially, reading
the notifications received from the transaction process(es) from a
concurrency-safe FIFO queue. Fortunately, the operation of signing is fast
enough that this may be done a thousand or more times per second.

The following caveats exist in regards to project transactions and immediate
availability of target files:

> - PyPI allows adding files to a project release at arbitrary later times.

Reviewer 3: What does this mean?

Author: Ernest mentioned this in https://github.com/python/peps/pull/1203#discussion_r336229332:
> PyPI allows for upload of additional files for a release at any point in the future

I can use his wording, if you prefer.

Reviewer: The original is clearer, though I can improve it a bit.
```suggestion
- PyPI allows permits additional  files to be uploaded to a project release at any point in the future.
```

- PyPI may want to make files available for download before all files of the
  project release have been uploaded.

> To solve this, PyPI MAY define variable transaction windows that mark a

Reviewer 3: I don't understand this paragraph at all. Maybe we should discuss offline.

transaction as finished even before all target files of a project release have
been uploaded. A variable transaction window SHOULD be chosen so that a client,
who downloads target files of a project release, is not left in an inconsistent

> state (with our without TUF metadata).

Reviewer 2: ```suggestion
state (with or without TUF metadata).
```

Furthermore, the transaction window MAY only apply to the generation of a TUF
consistent snapshot. That is, uploaded target files MAY become available for
download immediately after being uploaded and before TUF has generated a
consistent snapshot. If a user then downloads those target files in that short
time frame, the client (e.g. pip) SHOULD warn the user about missing TUF
protection and advice to abort download and try later.

At any rate, PyPI SHOULD use a `transaction log`__ to record project
transaction processes and the snapshot queue for auditing and to recover from
errors after a server failure.

__ https://en.wikipedia.org/wiki/Transaction_log

Cleaning up old metadata
------------------------

To avoid running out of disk space, by constantly producing new consistent

> snapshots, PyPI SHOULD regularly delete old consistent snapshots, i.e. metadata

Reviewer: ```suggestion
To avoid running out of disk space due to the constant production of new consistent
snapshots, PyPI SHOULD regularly delete older ones, i.e. metadata
```

and target files that were obsoleted some reasonable time in the past, such as
1 hour.

In order to preserve the latest consistent snapshot PyPI MAY use a
"mark-and-sweep" algorithm. That is, walk from the root of the latest
consistent snapshot, i.e. *timestamp* over *snapshot* over *targets* and
delegated targets until the target files, marking all visited files, and
delete all unmarked files. The last few consistent snapshots may be preserved
in a similar fashion.

Deleting a consistent snapshot will cause clients to see nothing except HTTP
404 responses to any request for a file within that consistent snapshot.
Clients SHOULD then retry their requests (as before) with the latest consistent
snapshot.

> Note that *root* metadata, even though versioned, is not part of any consistent

Reviewer 3: ```suggestion
Note that *root* metadata, versioned or not, is a part of all consistent
```

Author: Wait, what? No. :)

Author: I think we need to discuss this too.

> snapshot. PyPI MUST NOT delete old versions of *root* metadata. This guarantees

Reviewer 3: ```suggestion
snapshots. PyPI MUST NOT delete old versions of *root* metadata. This guarantees
```
